### DISCUSSION TITLE
Correct pattern of Generating Migrations

### DISCUSSION THREAD

**Discussion by Taklop** [prisma/prisma#12685] [Q&A] [answered]
I am using prisma with postgres.
I have a fresh DB, starting from scratch. I add a new schema to the `schema.prisma` file.
I run `prisma generate`. It succeeds, tells me I can start using the client in my code.

Next, I run `prisma migrate deploy`. It states `No migration found in prisma/migrations`. (This is technically true, there are no migration files that have been created?)

So I build, and run my app, then boom:

```
Error: 
Invalid `prisma.<TABLENAME>.findMany()` invocation:


  The table `public.TABLENAME` does not exist in the current database.
```

I check on prisma studio, and indeed, there is no table.

So, why is the table missing, if I have "run" the migration?

[In the docs](https://www.prisma.io/docs/concepts/components/prisma-migrate/db-push#choosing-db-push-or-prisma-migrate) it suggests that migrate is correct to use on CI, not db push. It also mentions that it will apply **pending** migrations. What is the workflow to generate these pending migrations?

If I run `npx prisma migrate dev`, it prompts me for a name, and generates the migrations in the folder as required!

So my question is **do you _have_ to run the dev migrations, before you run the deploy migrations?**

Is the flow to:

- Edit the schema file
- Run `npx prisma migrate dev`
  - (Ensure the generated migration folder is in VC)
- Deploy to CI CD
- Use the migrate deploy to apply those dev generated migrations

If this is the case, what is the purpose of the db push?

I just want to make sure I am following the correct flow 😄

**Comment by austincrim** [comment, marked as answer]
Hey @Taklop 👋 ,

> Is the flow to:
Edit the schema file
Run npx prisma migrate dev
    (Ensure the generated migration folder is in VC)
Deploy to CI CD
Use the migrate deploy to apply those dev generated migrations

This is correct and is the intended workflow for using Migrate. [Our docs page on Migrate](https://www.prisma.io/docs/concepts/components/prisma-migrate#getting-started-with-prisma-migrate) outlines a lot more info, if you're curious.

> If this is the case, what is the purpose of the db push?

[`prisma db push`](https://www.prisma.io/docs/concepts/components/prisma-migrate/db-push) does not generate any migration files and prioritizes the desired end state of a schema. It is mostly used for prototyping where a migration history is not necessary or wanted yet.

If I can clarify anything further, let me know.

**Reply by Jolg42** [reply]
Additionally, there is a page outlining the differences here with more details [Choosing db push or Prisma Migrate](https://www.prisma.io/docs/concepts/components/prisma-migrate/db-push#choosing-db-push-or-prisma-migrate)

**Reply by Taklop** [reply]
This all makes sense!

I guess I was confused that you had to first migrate with the "dev" flag to actually generate the migrations files.

As in, it's the same command, but with different flags. Dev to generate & apply, deploy to only apply.

This all makes sense to me now!

And I guess the pattern for people setting up a repo that contains prisma would be to pull it all, install deps etc, then `prisma db push` so they make their local DB the same as the "current state" of the schema?

**Reply by austincrim** [reply]
That is helpful feedback, thank you!

> And I guess the pattern for people setting up a repo that contains prisma would be to pull it all, install deps etc, then prisma db push so they make their local DB the same as the "current state" of the schema?

That could definitely be one use case. Here are a few points we make in that docs page that Joël linked above:

<blockquote>
db push works well if:

- You want to quickly prototype and iterate on schema design locally without the need to deploy these changes to other environments such as other developers, or staging and production environments.

- You are prioritizing reaching a desired end-state and not the changes or steps executed to reach that end-state (there is no way to preview changes made by db push)

- You do not need to control how schema changes impact data. There is no way to orchestrate schema and data migrations - if db push anticipates that changes will result in data loss, you can either accept data loss with the --accept-data-loss option or stop the process - there is no way to customize the changes.
</blockquote>

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
